ldap auth mode

cyberb opened the pull request:
### Purpose

Allows gui to authenticate against ldap using simple bind approach.

Additional config section:

    <authMode>ldap</authMode>
    <ldapServer>localhost</ldapServer>
    <ldapPort>389</ldapPort>
    <ldapBindDn>cn=%s,ou=users,dc=syncloud,dc=org</ldapBindDn>

Existing user/password authentication is preserved and can be used as before if authMode is not present or not equals to ldap.

### Testing

Few mising unit tests are added for existing auth.
No unit tests for ldap, although integration tests are available in Syncloud project:
https://github.com/syncloud/syncthing/blob/master/integration/verify.py#L99

## Authorship

Boris Rybalkin (ribalkin@gmail.com)

AudriusButkevicius commented:
So in principle this looks okay'ish to me other tham formatting issues etc. I would however like to propose for discussion (to see if people agree or disagree) hiding all the ldap settings behind an environment variable rather than config, this way we don't bleed ldap supports existance to the config which would probably be very rarely used anyway.nas in setting STLDAPSERVER=foo.com:1234 STLDAPSUBJECT=foo and checking if these env vars are set switching to ldap auth.

calmh commented on cmd/syncthing/gui_auth.go:
```diff
@@ -128,6 +110,45 @@ func basicAuthAndSessionMiddleware(cookieName string, cfg config.GUIConfiguratio
+func AuthSimple(username string, password string, configUser string, configPassword string) (bool, string) {
+
+    if username != configUser {
+        username = string(iso88591ToUTF8([]byte(username)))
```
This "might be UTF-8, might be ISO-8859-1" business with the username and password is a property of HTTP basic auth and affects auth over LDAP as much as built in auth. Hence it should be cleared up on some higher level.

calmh commented on lib/config/guiconfiguration.go:
```diff
@@ -26,6 +30,18 @@ type GUIConfiguration struct {
+}
+
+func (c GUIConfiguration) IsAuthModeLdap() bool {
+    return c.AuthMode == "ldap"
```
AuthMode here should be an enum of sorts and not a plain string I think.

AudriusButkevicius commented on lib/config/guiconfiguration.go:
```diff
@@ -12,15 +12,28 @@ import (
+const (
+	AuthModeSimple = "simple"
+	AuthModeLDAP   = "ldap"
+)
```
These consts probably should be their own types and enums, like FolderType et al.

AudriusButkevicius commented on cmd/syncthing/gui_auth.go:
```diff
@@ -110,43 +116,62 @@ func basicAuthAndSessionMiddleware(cookieName string, cfg config.GUIConfiguratio
-        log.Println(err)
-        return false
-    }
+	if cfg.IsLDAPTLSModeStartTSL() {
```
Instead of this IsModeX IsModeY I'd do
```
switch cfg.GetTLSMode() {
   case TlsModeFoo:
        connection = foo.connect()
   case TlsModeBar:
         connection = barr.dial() 
```

then you only need one accessor on the config object.

cyberb replied:
It is actually:
Plain:
  dial
Tls:
  dialTls
StartTls:
  dial
  startTls

I can do it like this if you want, just dial will be repeated :)

AudriusButkevicius commented on lib/config/guiconfiguration.go:
```diff
@@ -12,15 +12,28 @@ import (
+	LDAPPort                  int    `xml:"LDAPPort,omitempty" json:"LDAPPort"`
+	LDAPBindDn                string `xml:"LDAPBindDn,omitempty" json:"LDAPBindDn"`
+	LDAPTLSMode               string `xml:"LDAPTLSMode,omitempty" json:"LDAPTLSMode"`
+	LDAPInsecureSkipVerify    bool   `xml:"LDAPInsecureSkipVerify,attr" json:"LDAPInsecureSkipVerify" default:"false"`
```
Again, discussion point before we make any changes.

If we are starting to add LDAP support to syncthing, perhaps we should have a separate LDAPConfiguration section in the config, and then GUIAuthMode == LDAP uses the LDAPConfiguration, and then once we add some other non-GUI feature that uses LDAP, we don't have to add yet another LDAP configuration to some other part of the config.


/cc @calmh

calmh replied:
Yeah, I agree with giving this thing its own config struct/section, and perhaps combining server+port into just an address, we don't typically separate the parts in other places of the config.

AudriusButkevicius replied:
I think you pick one or the other?

calmh replied:
Maybe. Or we fall back to trying static if LDAP is down? Can we distinguish "down" from "denied" based on the bind result?

cyberb replied:
Technically Dial is up/down check and Bind is allowed/denied check and they are separate.

calmh replied:
Cool then it's easy to separate. So what's the desired outcome here? Assuming LDAP is configured,

- LDAP up, bind denied => denied
- LDAP down, static auth configured => follow static auth
- LDAP down, no static auth => denied (I guess? or allowed?)

cyberb replied:
I personally prefer no fallback. If i set auth mode to be ldap and it does not work i would like know right away as an admin, while fallback would waste my time.
I would say:
1. Auth mode is set to ldap do only ldap
2. Auth mode is set to static do only static
3. Auth mode not set do static if login and password set (config backward compatibility for now)

cyberb replied:
I am working on enums as suggested. Static is better, agree. Also extracting ldap config section. Will do another push soon.
1. Can you show me where the docs are?
2. Also I am still not sure how do you do deps. I had to do go get ./... but your TeamCity seems to not do that, so I had to include libs in vendor dir. Is this ok?

calmh replied:
Docs are at docs.syncthing.net, github.com/syncthing/docs. Dependencies need to be vendored, yes. We use gvt (https://github.com/FiloSottile/gvt). Essentially, `gvt fetch gopkg.in/whatever` and it'll add it to `vendor/` and update the manifest.

AudriusButkevicius commented on cmd/syncthing/gui_auth_test.go:
```diff
@@ -16,26 +16,24 @@ import (
-    if username != "user" {
-        t.Fatalf(username)
-    }
+
```
Spurious new line

AudriusButkevicius commented on cmd/syncthing/gui_auth.go:
```diff
@@ -110,43 +116,62 @@ func basicAuthAndSessionMiddleware(cookieName string, cfg config.GUIConfiguratio
+
+	configPasswordBytes := []byte(configPassword)
+	passwordBytes := []byte(password)
+	if err := bcrypt.CompareHashAndPassword(configPasswordBytes, passwordBytes); err != nil {
```
Can just be `return CompareAndHasPassword(...) == nil`

AudriusButkevicius commented on cmd/syncthing/gui_auth.go:
```diff
@@ -110,43 +116,62 @@ func basicAuthAndSessionMiddleware(cookieName string, cfg config.GUIConfiguratio
+}
+
+func AuthSimple(username string, password string, configUser string, configPassword string) bool {
+
```
blank line

AudriusButkevicius commented on cmd/syncthing/gui_auth.go:
```diff
@@ -110,43 +116,62 @@ func basicAuthAndSessionMiddleware(cookieName string, cfg config.GUIConfiguratio
+		return AuthLDAP(username, password, cfg)
+	} else {
+		return AuthSimple(username, password, cfg.User, cfg.Password)
+	}
```
Given auth is private (lowercase), no point for auth methods to be public.

AudriusButkevicius commented on cmd/syncthing/gui_auth.go:
```diff
@@ -110,43 +116,62 @@ func basicAuthAndSessionMiddleware(cookieName string, cfg config.GUIConfiguratio
-        return false
-    }
+func AuthLDAP(username string, password string, cfg config.GUIConfiguration) bool {
+
```
blank line

AudriusButkevicius commented on cmd/syncthing/gui_auth.go:
```diff
@@ -110,43 +116,62 @@ func basicAuthAndSessionMiddleware(cookieName string, cfg config.GUIConfiguratio
 
-    defer l.Close()
+	if err != nil {
+		log.Println(err)
```
We don't use log package, we have `l.{Debug/Info/Warn}ln` for this type of stuff.

cyberb commented:
1. Renamed all the ldap/Ldap to LDAP
2. Added LDAPTLSMode: nontls/tls/starttls (did not test tls/starttls, based on examples from: https://github.com/go-ldap/ldap/blob/master/ldap_test.go)
3. Not sure how to properly do enums in go, so added few consts
4. Not sure how to properly handle dependencies so added ldap lib into vendor libs (to make TeamCity happy)
5. Not sure how to run tests so removed test import cycle (to make TeamCity happy)

AudriusButkevicius commented on vendor/gopkg.in/ldap.v2/LICENSE:
```diff
@@ -0,0 +1,22 @@
+The MIT License (MIT)
```
I'd prefer that we used the github hosted version of the package. Given we use vendoring, and go modules just around the corner, using gopkg.in makes little sense.

calmh replied:
This thing also imports `gopkg.in/asn1-ber.v1` so while I agree in principle I suspect the fight is lost on this package

calmh commented on cmd/syncthing/gui_auth.go:
```diff
@@ -128,6 +116,64 @@ func basicAuthAndSessionMiddleware(cookieName string, cfg config.GUIConfiguratio
+	}
+}
+
+func authSimple(username string, password string, configUser string, configPassword string) bool {
```
It's not strictly required for this PR, but the auth functions that have the same signature (or could have) could be types that implement an interface, with something that creates the right one based on the config. It's a bit boilerplate-ish and can be easily added after the fact so skip that for now if you think it's easier.

Not doing any further style nitpicking as Audrius is already all over it. :)

cyberb commented:
1. Extracted enums (ldap configuration, auth type)
2. This time I used gvt fetch gopkg.in/ldap.v2

cyberb commented:
Let me check correct logger

AudriusButkevicius commented on lib/config/ldaptlsmode.go:
```diff
@@ -0,0 +1,46 @@
+
+func (t LDAPTLSMode) String() string {
+	switch t {
+	case LDAPTLSModeNoTLS:
```
Perhaps this should be plain not notls?

AudriusButkevicius commented on cmd/syncthing/gui_auth.go:
```diff
@@ -128,6 +116,60 @@ func basicAuthAndSessionMiddleware(cookieName string, cfg config.GUIConfiguratio
+}
+
+func authStatic(username string, password string, configUser string, configPassword string) bool {
+
```
Blank line

AudriusButkevicius commented on lib/config/authmode.go:
```diff
@@ -0,0 +1,41 @@
+
+const (
+	AuthModeStatic AuthMode = iota // default is static
+	AuthModeLDAP
```
I would add a None mode and stop checking for username and password from being set, and do a migration in the config setting auth mode.

This would also need fixes in the UI.

AudriusButkevicius replied:
@calmh wdyt?

calmh replied:
That would seem clean but if that were the default it would mean existing things that just set username and password would result in no auth which is bad.

AudriusButkevicius commented on cmd/syncthing/gui_auth.go:
```diff
@@ -128,6 +116,60 @@ func basicAuthAndSessionMiddleware(cookieName string, cfg config.GUIConfiguratio
+
+func authLDAP(username string, password string, cfg config.LDAPConfiguration) bool {
+
+	address := cfg.LDAPAddress
```
Blank line

AudriusButkevicius commented on cmd/syncthing/gui_auth.go:
```diff
@@ -128,6 +116,60 @@ func basicAuthAndSessionMiddleware(cookieName string, cfg config.GUIConfiguratio
+
+	configPasswordBytes := []byte(configPassword)
+	passwordBytes := []byte(password)
+	return bcrypt.CompareHashAndPassword(configPasswordBytes, passwordBytes) == nil
```
This might have been this way, but this just hit me. 

Perhaps move the username check after the hash check to prevent timing attacks on username probing.

calmh replied:
Currently both username and password is checked, and failure on either results in a short random delay before returning an error.

AudriusButkevicius commented on cmd/syncthing/gui_auth.go:
```diff
@@ -128,6 +116,60 @@ func basicAuthAndSessionMiddleware(cookieName string, cfg config.GUIConfiguratio
+	}
+
+	if err != nil {
+		log.Println(err)
```
Please remove all usages of the log package and use the `l` logger instance.

AudriusButkevicius commented on lib/config/guiconfiguration.go:
```diff
@@ -13,17 +13,27 @@ import (
+	User                      string            `xml:"user,omitempty" json:"user"`
+	Password                  string            `xml:"password,omitempty" json:"password"`
+	AuthMode                  AuthMode          `xml:"authMode,omitempty" json:"authMode"`
+	LDAPAuth                  LDAPConfiguration `xml:"ldapAuth,omitempty" json:"ldapAuth"`
```
I would move LDAPConfigurarion as LDAPConfiguration to Config instead of GUIConfiguration.

calmh commented on lib/config/config.go:
```diff
@@ -125,17 +125,18 @@ func ReadJSON(r io.Reader, myID protocol.DeviceID) (Configuration, error) {
+	Folders           []FolderConfiguration `xml:"folder" json:"folders"`
+	Devices           []DeviceConfiguration `xml:"device" json:"devices"`
+	GUI               GUIConfiguration      `xml:"gui" json:"gui"`
+	LDAPConfiguration LDAPConfiguration     `xml:"ldapConfiguration" json:"ldapConfiguration"`
```
Drop the `Configuration` part of the JSON/XML names I think.

AudriusButkevicius replied:
Also, potentially for the member name itself? GUIConfiguration is called GUI.

calmh replied:
Aye

AudriusButkevicius commented on cmd/syncthing/gui_auth.go:
```diff
@@ -128,6 +115,58 @@ func basicAuthAndSessionMiddleware(cookieName string, cfg config.GUIConfiguratio
+	}
+
+	if err != nil {
+		l.Warnln(err)
```
If this is `connection reset by peer` error or alike, perhaps it would make sense adding context to it, something like
```l.Warnln("ldap.Dial", err)```

AudriusButkevicius commented on cmd/syncthing/gui_auth.go:
```diff
@@ -128,6 +115,58 @@ func basicAuthAndSessionMiddleware(cookieName string, cfg config.GUIConfiguratio
+	if cfg.LDAPTLSMode == config.LDAPTLSModeStartTLS {
+		err = connection.StartTLS(&tls.Config{InsecureSkipVerify: cfg.LDAPInsecureSkipVerify})
+		if err != nil {
+			l.Warnln(err)
```
ditto about context

AudriusButkevicius commented on cmd/syncthing/gui_auth.go:
```diff
@@ -128,6 +115,58 @@ func basicAuthAndSessionMiddleware(cookieName string, cfg config.GUIConfiguratio
+
+	err = connection.Bind(fmt.Sprintf(cfg.LDAPBindDn, username), password)
+	if err != nil {
+		l.Warnln(err)
```
ditto about context

AudriusButkevicius commented on lib/config/guiconfiguration.go:
```diff
@@ -13,17 +13,26 @@ import (
+}
+
+func (c GUIConfiguration) IsAuthModeStatic() bool {
+	return (len(c.User) > 0 && len(c.Password) > 0)
```
no need for the outer braces

AudriusButkevicius commented on lib/config/ldaptlsmode.go:
```diff
@@ -0,0 +1,46 @@
+func (t LDAPTLSMode) String() string {
+	switch t {
+	case LDAPTLSModePlain:
+		return "plain"
```
Now that I think about it, it's not plain tls, it's simply not tls at all, so perhaps this should not be LDAPTLSMode, but LDAPTransport or LDAPProtocol or something alike. Or the mode should be none, (as in non-tls).

Sorry for distracting you here earlier.

cyberb replied:
Lets do LDAPTransport with LDAPTransportPlain, LDAPTransportTLS and LDAPTransportStartTLS?

AudriusButkevicius commented:
Minus the nitpicks, this looks good to me.

calmh commented on cmd/syncthing/gui_auth.go:
```diff
@@ -124,34 +124,30 @@ func auth(username string, password string, guiCfg config.GUIConfiguration, ldap
-
 	configPasswordBytes := []byte(configPassword)
 	passwordBytes := []byte(password)
-	return bcrypt.CompareHashAndPassword(configPasswordBytes, passwordBytes) == nil
```
Nice

calmh commented on lib/config/ldapconfiguration.go:
```diff
@@ -7,10 +7,10 @@
-	LDAPBindDn             string      `xml:"bindDn,omitempty" json:"bindDn"`
-	LDAPTLSMode            LDAPTLSMode `xml:"tlsMode,omitempty" json:"tlsMode"`
-	LDAPInsecureSkipVerify bool        `xml:"insecureSkipVerify,omitempty" json:"insecureSkipVerify" default:"false"`
+	LDAPAddress            string        `xml:"address,omitempty" json:"addresd"`
```
So, given we are now inside an LDAP struct we can skip the LDAP prefix here as well. And capitalize DN in BindDN.

cyberb commented:
Do you have any more comments?

AudriusButkevicius commented:
Oh, and thanks for your patience.
